### ISSUE TITLE
AI-优化方向

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9554]
AI-优化方向<br>
<br>
https://ift.tt/tEfdy2v<br>
<br>
Ph.D. Candidate at HUST<br>
<br>

<div><div><div><div><div><section><p>本文主要介绍优化AI的主要方向，包括三个方向:<strong>计算</strong>、<strong>网络</strong>和<strong>存储</strong>。重点分析两个方向:<strong>并行训练优化</strong>和<strong>GPU不足</strong>(<strong>大batch-size</strong>问题)。</p>
<h2>1. Computation / Training</h2>
<h3>1.1 算法、编译优化</h3>
<h3>1.2 训练优化</h3>
<ul>
<li><p><a href="https://arxiv.org/pdf/1805.08899.pdf">Echo(2018.5)</a>:compiler-based optimization scheme to ensure each feature map recomputation to be effective and efficient</p></li>
<li><p><a href="https://arxiv.org/pdf/1807.08887.pdf">Tofu(EuroSys 19)</a>: 自动划分模型并行，减少通信开销和显存占用</p></li>
<li><p><a href="https://arxiv.org/pdf/1910.02054.pdf">ZeRO(2019.10)</a>: 主要是一些优化显存占用的方法，将模型、梯度划分到数据并行的卡上</p></li>
<li>
<a href="https://cs.stanford.edu/~matei/papers/2019/sosp_pipedream.pdf">PipeDream(SOSP 19)</a><ul>
<li>自动划分最优的流水并行策略，保留多个版本的权重，通过调度可以填满流水线</li>
<li>通过weight stashing避免权重更新的staleness</li>
<li>pipeline能够减少通信量，也能进行计算通信的重叠</li>
</ul>
</li>
<li>
<a href="http://learningsys.org/sosp19/assets/papers/24_CameraReadySubmission_paper.pdf">(SOSP 19 workshop)</a>: a number of techniques that leverage cheap host memory to reduce GPU memory consumption during training in TensorFlow</li>
</ul>
<h3>1.3 框架开发和设计</h3>
<h3>1.4 具体应用</h3>
<ul>
<li>Computation Graph<ul><li>
<a href="https://cs.stanford.edu/~padon/taso-sosp19.pdf">TASO(SOSP 19)</a>: 第一个自动化的计算图优化，只考虑了forward的优化<ul>
<li>Graph Substitution Generator:根据Operator Specifications枚举自动生成所有可能的图替换</li>
<li>Graph Substitution Verifier:验证得到的图替换是正确的</li>
<li>去掉冗余的图替换</li>
<li>Graph Substitution &amp; Data Layout Joint Optimizer:对计算图进行优化，通过枚举所有的layout和图替换，找出有优化效果的替换</li>
</ul>
</li></ul>
</li>
<li>AutoML<ul>
<li>
<a href="https://arxiv.org/pdf/1810.05934.pdf">ASHA(2018.10)</a>: exploits parallelism and aggressive early-stopping to tackle large-scale hyperparameter optimization problems</li>
<li>
<a href="https://arxiv.org/pdf/1807.05118.pdf">Tune(ICML 18 workshop)</a>: a unified framework for model selection and training that provides a narrow-waist interface between training scripts and search algorithms</li>
</ul>
</li>
<li>性能预测<ul>
<li>
<a href="https://arxiv.org/pdf/2006.03318.pdf">Daydream(2020.6)</a>: Daydream models DNN execution with a fine-grained dependency graph based on low-level traces collected by CUPTI, and predicts runtime by simulating execution based on the dependency graph. 能够模拟大多数优化技巧</li>
<li>
<a href="https://arxiv.org/pdf/2002.06790.pdf">Offline Profiling(2020.2)</a>: uses operation-level profiling and dataflow based simulation估计速度</li>
</ul>
</li>
<li>Prediction Serving System<ul>
<li>
<a href="https://www.usenix.org/system/files/osdi18-lee.pdf">PRETZEL(OSDI 18)</a><ul>
<li>部署系统是一个黑箱，对可使用的优化技巧有限制，而且在其他方面也有很多限制，比如部署的服务只有在频繁使用时才能盈利</li>
<li>A set of design principles for white box model serving allowing pipelines to be optimized for inference and to share resources</li>
</ul>
</li>
<li>
<a href="https://homes.cs.washington.edu/~arvind/papers/nexus.pdf">NEXUS(SOSP 19)</a><ul>
<li>In order to realize the promise of very low-cost processing made by accelerators such as GPUs, it is essential to run them at sustained high utilization.</li>
<li>GPU集群细粒度调度, reasoning about groups of DNN invocations that need to be coscheduled,and moving from the conventional whole-DNN execution model to executing fragments of DNNs</li>
</ul>
</li>
<li>
<a href="https://dl.acm.org/doi/pdf/10.1145/3341301.3359654">parity models(SOSP 19)</a><ul>
<li>Predication serving systems are prone to slowdowns and failures that inflate tail latency</li>
<li>A parity model is a neural network trained to transform erasure-coded queries into a form that enables a decoder to reconstruct slow or failed predictions</li>
</ul>
</li>
</ul>
</li>
<li>straggler worker<ul>
<li>Gradient Encoding(ICML 17): 每张卡计算多个梯度，这样对整个系统来说，任意一部分机器都能够获得全局梯度的和</li>
<li>
<a href="http://learningsys.org/sosp19/assets/papers/18_CameraReadySubmission_CameraReady_FlexEnt.pdf">FlexEnt(SOSP 19 workshop)</a>: identify the entropy present in a training batch using a set of “feature filters” and replicates only a fraction of the training data on a predetermined set of “shadow workers”</li>
</ul>
</li>
</ul>
<h2>2. communication</h2>
<h3>2.1 Algorithm Optimization</h3>
<ul>
<li>Large Batch Training<ul>
<li>LR Scaling Rules: Linear/Sqrt</li>
<li>Warmup Scheme: Gradual Warmup / LEGW</li>
<li>LARS / LAMB</li>
</ul>
</li>
<li>Periodic Communication</li>
<li>Async Compensation</li>
</ul>
<h3>2.2 Compression</h3>
<ul>
<li>quantization</li>
<li>Sparsification</li>
<li>Decomposition</li>
</ul>
<h3>2.3 Scheduler</h3>
<h3>2.4 Logical Architecture</h3>
<h3>2.5 communication library</h3>
<h3>2.6 Software &amp; Hardware/System co-design</h3>
<h2>3. Stoage</h2>
<h3>3.1 Cluster Scheduler</h3>
<h3>3.2 显存管理</h3>
<ul>
<li><p><a href="http://learningsys.org/sosp19/assets/papers/9_CameraReadySubmission_TF_allocator_sosp19_camera_ready.pdf">PRmalloc(SOSP 19 workshop)</a>: leverages the domainspecific knowledge to better cache and reuse large memory blocks, cutting down the system overhead of memory allocation and saving memory footprint significantly</p></li>
<li><p><a href="http://learningsys.org/sosp19/assets/papers/15_CameraReadySubmission_camera-ready.pdf">Intelligent Store(SOSP 19 workshop)</a>: optimization of the data path with zero-copy messaging, and avoiding suprious copying and locking.</p></li>
</ul>
<h3>3.3 异构集群</h3>
<ul><li>
<a href="http://learningsys.org/sosp19/assets/papers/7_CameraReadySubmission_AI_system_camera_ready_v1.pdf">OverSeer(SOSP 19 workshop)</a>: a topologyand hardware-agnostic end-to-end system, that uses a novel partitioning algorithm to efficiently utilize different computational capabilities present in heterogeneous devices.</li></ul>
<h3>3.4 容量问题</h3>
<ul><li>Behemoth、FlashNeuron(Fast 21)</li></ul>
<h2>4. 重点分析</h2>
<h3>4.1 PipeDream: Generalized Pipeline Parallelism for DNN Training</h3>
<h3>4.2 FlashNeuron: SSD-Enabled Large-Batch Training of Very Deep Neural Networks</h3>
<p><img alt="image-20210401093048459" src="https://emperorlu.github.io/photos/image-20210401093048459.png"></p>
<ul>
<li>
<p>Memory Manager</p>
<ul>
<li>
<strong>管理GPU内存空间(分配和回收)、避免碎片化、预取、tensor压缩、降低精度</strong>等等</li>
<li>自定义内存分配器首先预留整个GPU内存空间，并自行管理内存分配/重分配<ul>
<li>当 i)一个张量在正向传播过程中首次创建或 ii)一个卸载的张量在向后传递过程中从SSD预取到内存时，就会分配张量</li>
<li>另一方面，当i)一个张量完全从内存卸载到SSD，或者 ii)一个张量在迭代期间不再被任何层使用时，张量被重新分配</li>
<li>引用计数for寿命(PyTorch自带)</li>
</ul>
</li>
<li>避免碎片化<ul>
<li>如果从一开始就为所有张量分配内存地址，由于只有一些张量被卸载到ssd上，可能会出现严重的内存碎片，从而有效地在GPU物理内存地址空间中制造洞</li>
<li>为了避免这个问题，从内存地址空间的最低端分配内存常驻(即未卸载)张量，从内存地址空间的最高端分配临时(即已卸载)张量。由于临时数据在向前传递期间的生命周期很短，因此只使用了很小一部分内存地址空间</li>
</ul>
</li>
<li>Managing Offloading and Prefetching</li>
<li>CSR压缩仅适用于ReLU的输出张量(43%至75%稀疏度)，由于张量是一个多维矩阵，我们把张量转换成一个二维矩阵，它的列是128个条目。将存储每个元素的列索引的向量(通常称为JA向量)替换为一组bits向量，其中每个bit向量表示一行的一组非零元素</li>
</ul>
<p><img alt="crc" src="https://emperorlu.github.io/photos/crc.png"><img alt="CSR" src="https://emperorlu.github.io/photos/CSR.png"></p>
<p>大小a = 原大小b - 非零个数m * 8 + 总数n * 1</p>
<p>a &lt; b 需要 m &gt; 1/8 * n，即1/8以上都是零，能够减少空间，ReLU的输出张量大部分都满足</p>
<ul><li>对卸载张量使用半精度浮点数(FP16)<ul>
<li>神经网络可以容忍一定程度的精度损失，而不会显著降低最终模型的精度</li>
<li>在前向传播期间，Memory Manager首先将卸载的张量转换为FP16格式(从FP32)，然后将它们存储在SSD中。稍后，在向后传播过程中，卸载的FP16张量将被预取、填充到FP32</li>
</ul>
</li></ul>
</li>
<li><p>Offloading Scheduler</p></li>
<li>
<p>Peer-to-Peer Direct Storage Access</p>
<ul>
<li><p>P2P-DSA建立在 GDRCopy 和 SPDK 的基础上，GDRCopy是基于NVIDIA GPUDirect技术的GPU内存快速拷贝库，该技术可以让GPU内存直接被其他PCIe外设备访问</p></li>
<li><p>为了维护卸载到ssd的每个张量元数据，P2P-DSA包含一个元数据表，该表由一个long LBA值和一个boolean组成，用于检查I/O完成情况</p></li>
<li>
<p>流程:</p>
<ul>
<li><p><strong>1.</strong>P2P-DSA从传输请求中获取索引、缓冲区和方向(写)信息</p></li>
<li><p><strong>2.</strong>调用逻辑块地址(LBA)分配器来在单个SSD设备或多个SSD设备(当使用多个SSD块提高卸载/预取带宽)上分配一组连续的，从LBA分配器分配的第一个块的LBA在元数据表的适当位置更新</p></li>
<li>
<p><strong>3.</strong>P2P-DSA为每个逻辑块创建一个命令，然后将其放入命令队列。这里的NVMe命令包含 i)源地址(GPU内存地址由GPUDirect转换为PCIe总线地址)和 ii)设备地址(使用元数据表中的LBA计算)</p>
<p><img alt="image-20210401100112556" src="https://emperorlu.github.io/photos/image-20210401100112556.png"></p>
<p>update:</p>
</li>
<li><p><strong>4.</strong>命令队列中的命令被获取并发给NVMe SSD，只要NVMe设备提交队列有空间</p></li>
<li>
<p><strong>5.</strong>NVMe SSD设备将执行这些请求，并在SSD设备与GPU之间直接进行数据传输。稍后，这些转移请求将被完成，并且它们的状态将在NVMe设备完成队列</p>
<p>再次update</p>
</li>
<li><p><strong>6.</strong>清除完成队列并通过设置相应的完成位来更新元数据表</p></li>
<li><p><strong>7.</strong>如果应用程序调用done，对于已经卸载的张量，它将返回true</p></li>
<li><p>反向传播(从SSD预取数据到GPU内存)的操作与此类似，不同的是: i)从元数据表中读取LBAs，而不是分配LBAs; ii)发出读命令，而不是写命令。在卸载和预取情况下，大多数数据访问都是顺序访问，这在吞吐量和持久时间上比随机访问更有利</p></li>
</ul>
</li>
</ul>
</li>
</ul>
<p><img alt="image-20210401103751948" src="https://emperorlu.github.io/photos/image-20210401103751948.png"></p>
<ul><li>
<p>优化思考</p>
<ul>
<li><p>效率是否高？性能如何？:测试验证</p></li>
<li><p>选择卸载和预取策略能否优化？</p></li>
<li><p>并行化处理，使用更高性能的NVM</p></li>
<li><p>如何更好的分配和使用GPU、NVMe</p></li>
</ul>
</li></ul></section></div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Emperorlu's Site <br>
<br>
June 25, 2023 at 11:34AM
